pkg/services/ngalert/ngalert.go

```diff
@@ -106,31 +109,38 @@ func (ng *AlertNG) init() error {
 
 	// Let's make sure we're able to complete an initial sync of Alertmanagers before we start the alerting components.
 	if err := ng.MultiOrgAlertmanager.LoadAndSyncAlertmanagersForOrgs(context.Background()); err != nil {
-		return err
-	}
-
-	schedCfg := schedule.SchedulerCfg{
-		C:                       clock.New(),
-		BaseInterval:            ng.Cfg.UnifiedAlerting.BaseInterval,
-		Logger:                  ng.Log,
-		MaxAttempts:             ng.Cfg.UnifiedAlerting.MaxAttempts,
-		Evaluator:               eval.NewEvaluator(ng.Cfg, ng.Log, ng.DataSourceCache, ng.SecretsService),
-		InstanceStore:           store,
-		RuleStore:               store,
-		AdminConfigStore:        store,
-		OrgStore:                store,
-		MultiOrgNotifier:        ng.MultiOrgAlertmanager,
-		Metrics:                 ng.Metrics.GetSchedulerMetrics(),
-		AdminConfigPollInterval: ng.Cfg.UnifiedAlerting.AdminConfigPollInterval,
-		DisabledOrgs:            ng.Cfg.UnifiedAlerting.DisabledOrgs,
-		MinRuleInterval:         ng.Cfg.UnifiedAlerting.MinInterval,
+		return fmt.Errorf("failed to initialize alerting because notifier manager fail to warm up: %w", err)
 	}
 
 	appUrl, err := url.Parse(ng.Cfg.AppURL)
 	if err != nil {
 		ng.Log.Error("Failed to parse application URL. Continue without it.", "error", err)
 		appUrl = nil
 	}
+
+	clk := clock.New()
+
+	dispatcher := sender.NewDispatcher(ng.MultiOrgAlertmanager, store, clk, appUrl, ng.Cfg.UnifiedAlerting.DisabledOrgs, ng.Cfg.UnifiedAlerting.AdminConfigPollInterval)
+
+	if err := dispatcher.SyncAndApplyConfigFromDatabase(); err != nil {
```

Comment:
I wrote a bunch of comments in `sender.go` below about where to do this Sync before seeing that it's done here after construction and before calling Run. I think the API would be simpler and less error prone if we didn't require the programmer to make this additional call when we know what the right thing to do is.

Reply:
After more thinking, I decided to address it in part-2 which will refactor tests. Currently, this small change could cause problems in scheduler's unit tests because they use the real dispatcher.

Reply:
Sounds good. I'm also fine with keeping the initial sync separate if that's consistent with how we configure all the services in this package.